src/Storages/IStorageDataLake.h

```diff
@@ -0,0 +1,229 @@
+
+        auto new_configuration = getAdjustedS3Configuration(context_, base_configuration, log);
+
+        if (columns_.empty())
+        {
+            columns_ = StorageS3::getTableStructureFromData(
+                new_configuration, /*distributed processing*/ false, format_settings_, context_, nullptr);
+            storage_metadata.setColumns(columns_);
+        }
+        else
+            storage_metadata.setColumns(columns_);
+
+        storage_metadata.setConstraints(constraints_);
+        storage_metadata.setComment(comment);
+        setInMemoryMetadata(storage_metadata);
+
+        s3engine = std::make_shared<StorageS3>(
+            new_configuration,
+            table_id_,
+            columns_,
+            constraints_,
+            comment,
+            context_,
+            format_settings_,
+            /* distributed_processing_ */ false,
+            nullptr);
+    }
+
+    static constexpr auto name = Name::name;
+    String getName() const override { return name; }
+
+    // Reads latest version of Lake Table
+    Pipe read(
+        const Names & column_names,
+        const StorageSnapshotPtr & storage_snapshot,
+        SelectQueryInfo & query_info,
+        ContextPtr context,
+        QueryProcessingStage::Enum processed_stage,
+        size_t max_block_size,
+        size_t num_streams) override
+    {
+        StorageS3::updateS3Configuration(context, base_configuration);
+
+        return s3engine->read(column_names, storage_snapshot, query_info, context, processed_stage, max_block_size, num_streams);
+    }
+
+    static ColumnsDescription getTableStructureFromData(
+        StorageS3::Configuration & configuration, const std::optional<FormatSettings> & format_settings, ContextPtr ctx)
+    {
+        StorageS3::updateS3Configuration(ctx, configuration);
+        auto new_configuration = getAdjustedS3Configuration(ctx, configuration, &Poco::Logger::get("Storage" + String(name)));
+        return StorageS3::getTableStructureFromData(
+            new_configuration, /*distributed processing*/ false, format_settings, ctx, /*object_infos*/ nullptr);
+    }
+
+    static StorageS3::Configuration
+    getAdjustedS3Configuration(const ContextPtr & context, StorageS3::Configuration & configuration, Poco::Logger * log)
+    {
+        MetaParser parser{configuration, context};
+
+        auto keys = parser.getFiles();
+        String new_uri = std::filesystem::path(configuration.url.uri.toString()) / Name::data_directory_prefix
+            / MetaParser::generateQueryFromKeys(keys, configuration.format);
+
+        StorageS3::Configuration new_configuration(configuration);
+        new_configuration.url = S3::URI(new_uri);
+
+        LOG_DEBUG(log, "Table path: {}, new uri: {}", configuration.url.key, new_uri);
+
+        return new_configuration;
+    }
+
+    static void processNamedCollectionResult(StorageS3::Configuration & configuration, const NamedCollection & collection)
```

Comment:
this method is the same as in StorageS3? Why not reuse then?

Reply:
It's difference, in StorageS3, it may specify a file name, but this is not.

Reply:
not possible - is it because we need to parse arguements in a certain way different from for example StorageS3? E.g. we do not need `compression_method` and `format` arguments as the first one is not supported and the second one should be Parquet always? Then seems like we can allow to parse the arguments as StorageS3 usually does, and then check the result having just a base class of `S3Storage::Configuration : public StatelessTableEngineConfiguration` (from Storages/StorageConfiguration.h) which has format and compression methods, and throw exception or rewrite them if needed. WDYT? E.g. as I understand it is enough for us to have pointer to base StatelessTableEngineConfiguration here to resolve the differences.